dm/master/openapi_test.go

```diff
@@ -207,3 +218,177 @@ func (t *openAPISuite) TestSourceAPI(c *check.C) {
+		c.Assert(ha.KeepAlive(ctx, s.etcdClient, workerName, keepAliveTTL), check.IsNil)
+	}(ctx1, workerName1)
+	// wait worker ready
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		w := s.scheduler.GetWorkerBySource(source1.SourceName)
+		return w != nil
+	}), check.IsTrue)
+	// mock worker rpc
+	mockWorkerClient := pbmock.NewMockWorkerClient(ctrl)
+	workerReq := &workerrpc.Request{
+		Type:        workerrpc.CmdQueryStatus,
+		QueryStatus: &pb.QueryStatusRequest{Name: ""},
+	}
+	mockWorkerClientForOpenAPI(mockWorkerClient, "", source1.SourceName, workerName1, workerReq)
+	s.scheduler.SetWorkerClientForTest(workerName1, newMockRPCClient(mockWorkerClient))
+
+	// get source status again,source should be bounded by worker1
+	result3 := testutil.NewRequest().Get(source1StatusURL).Go(t.testT, s.echo)
+	c.Assert(result3.Code(), check.Equals, http.StatusOK)
+	var resultSourceStatus1 openapi.SourceStatus
+	err = result3.UnmarshalBodyToObject(&resultSourceStatus1)
+	c.Assert(err, check.IsNil)
+	c.Assert(resultSourceStatus1.WorkerName, check.Equals, workerName1) // worker1 is bound
+	c.Assert(resultSourceStatus1.EnableRelay, check.Equals, true)
+
+	// mock start relay
+	startRelayReq := pb.OperateRelayRequest{}
+	mockWorkerClientForOpenAPI(mockWorkerClient, "", source1.SourceName, workerName1, startRelayReq)
+	startRelayURL := fmt.Sprintf("%s/start-relay", source1URL)
+	openAPIStartRelayReq := openapi.StartRelayRequest{WorkerName: workerName1}
+	result4 := testutil.NewRequest().Patch(startRelayURL).WithJsonBody(openAPIStartRelayReq).Go(t.testT, s.echo)
+	// check http status code
+	c.Assert(result4.Code(), check.Equals, http.StatusOK)
+
+	// get source status again, relay status should not be nil
+	result5 := testutil.NewRequest().Get(source1StatusURL).Go(t.testT, s.echo)
+	c.Assert(result5.Code(), check.Equals, http.StatusOK)
+	var resultSourceStatus2 openapi.SourceStatus
+	err = result5.UnmarshalBodyToObject(&resultSourceStatus2)
+	c.Assert(err, check.IsNil)
+	c.Assert(resultSourceStatus2.RelayStatus.Stage, check.Equals, pb.Stage_Running.String())
+
+	// test stop relay
+	stopRelayURL := fmt.Sprintf("%s/stop-relay", source1URL)
+	workerNameReq := openapi.WorkerNameRequest{WorkerName: workerName1}
+	result6 := testutil.NewRequest().Patch(stopRelayURL).WithJsonBody(workerNameReq).Go(t.testT, s.echo)
+	c.Assert(result6.Code(), check.Equals, http.StatusOK)
+}
+
+func mockWorkerClientForOpenAPI(
+	mockWorkerClient *pbmock.MockWorkerClient, taskName, sourceName, workerName string, masterReq interface{}) {
+	var expect pb.Stage
+	switch req := masterReq.(type) {
+	case *pb.StartTaskRequest, *pb.UpdateTaskRequest:
+		expect = pb.Stage_Running
+	case *pb.OperateTaskRequest:
+		switch req.Op {
+		case pb.TaskOp_Resume:
+			expect = pb.Stage_Running
+		case pb.TaskOp_Pause:
+			expect = pb.Stage_Paused
+		case pb.TaskOp_Stop:
+			expect = pb.Stage_Stopped
+		}
+	case *pb.OperateWorkerRelayRequest:
```

Comment:
start-relay and stop-relay are using OperateRelayRequest

Reply:
There is actually no need to mock rpc anymore, as start-realy just puts relay stage cfg into ETCD